crates/vm/derive/src/tco.rs

```diff
@@ -0,0 +1,128 @@
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::{parse_macro_input, ItemFn};
+
+/// Implementation of the TCO handler generation logic.
+/// This is called from the proc macro attribute in lib.rs.
+pub fn tco_impl(item: TokenStream) -> TokenStream {
+    // Parse the input function
+    let input_fn = parse_macro_input!(item as ItemFn);
+
+    // Extract information from the function
+    let fn_name = &input_fn.sig.ident;
+    let generics = &input_fn.sig.generics;
+    let where_clause = &generics.where_clause;
+
+    // Extract the first two generic type parameters (F and CTX)
+    let (f_type, ctx_type) = extract_f_and_ctx_types(generics);
+    // Derive new function name:
+    // If original ends with `_impl`, replace with `_tco_handler`, else append suffix.
+    let new_name_str = fn_name
+        .to_string()
+        .strip_suffix("_impl")
+        .map(|base| format!("{base}_tco_handler"))
+        .unwrap_or_else(|| format!("{fn_name}_tco_handler"));
+    let handler_name = format_ident!("{}", new_name_str);
+
+    // Build the generic parameters for the handler, preserving all original generics
+    let handler_generics = generics.clone();
+
+    // Build the function call with all the generics
+    let generic_args = build_generic_args(generics);
+    let execute_call = if generic_args.is_empty() {
+        quote! { #fn_name(pre_compute, exec_state) }
+    } else {
+        quote! { #fn_name::<#(#generic_args),*>(pre_compute, exec_state) }
+    };
+
+    // Generate the TCO handler function
+    let handler_fn = quote! {
+        #[inline(never)]
+        unsafe fn #handler_name #handler_generics (
+            interpreter: &::openvm_circuit::arch::interpreter::InterpretedInstance<#f_type, #ctx_type>,
```

Comment:
not sure if it matters but maybe we just pass like a slimmed down instruction table that contains a function/mapping from pc -> (data, handler) instead of a reference to this whole struct

Reply:
yea will do with the register pinning